test_runner/src/main/kotlin/ftl/json/SavedMatrix.kt

```diff
@@ -3,118 +3,116 @@ package ftl.json
+)
+
+fun createSavedMatrix(testMatrix: TestMatrix) = defaultSavedMatrix().updateWithMatrix(testMatrix)
 
-    var billableVirtualMinutes: Long = 0
-        private set
-    var billablePhysicalMinutes: Long = 0
-        private set
-    var outcome: String = ""
-        private set
-    var outcomeDetails: String = ""
-        private set
-    val clientDetails = matrix.clientInfo?.clientInfoDetails
-        ?.map { it.key to it.value }
-        ?.toMap()
-
-    init {
-        if (this.state == FINISHED) finished(matrix)
-    }
-
-    // ExitCodeFromRollupOutcome - https://github.com/bootstraponline/gcloud_cli/blob/137d864acd5928baf25434cf59b0225c4d1f9319/google-cloud-sdk/lib/googlecloudsdk/api_lib/firebase/test/exit_code.py#L46
-    fun failed(): Boolean {
-        // outcome of the test execution.
-        // skipped means unsupported environment
-        return when (outcome) {
-            failure -> true
-            skipped -> true
-            inconclusive -> true
-            else -> false
-        }
-    }
-
-    /** return true if the content changed **/
-    fun update(matrix: TestMatrix): Boolean {
-        val newState = matrix.state
-        val newLink = matrix.webLink()
-        val changedState = state != newState
-        val changedLink = webLink != newLink
+fun SavedMatrix.canceledByUser() = outcomeDetails == ABORTED_BY_USER_MESSAGE
 
-        if (changedState) {
-            updateState(newState, matrix)
-        }
+fun SavedMatrix.infrastructureFail() = outcomeDetails == INFRASTRUCTURE_FAILURE_MESSAGE
 
-        if (changedLink) {
-            this.webLink = newLink
-        }
+fun SavedMatrix.incompatibleFail() = outcomeDetails in arrayOf(
+    INCOMPATIBLE_APP_VERSION_MESSAGE,
+    INCOMPATIBLE_ARCHITECTURE_MESSAGE,
+    INCOMPATIBLE_DEVICE_MESSAGE
+)
+
+fun SavedMatrix.isFailed() = when (outcome) {
+    failure -> true
+    skipped -> true
+    inconclusive -> true
+    else -> false
+}
 
-        return changedState || changedLink
-    }
+fun SavedMatrix.needsUpdate(newMatrix: TestMatrix): Boolean {
+    val newState = newMatrix.state
+    val newLink = newMatrix.webLink()
+    val changedState = state != newState
+    val changedLink = webLink != newLink
+    return (changedState || changedLink)
+}
 
-    private fun updateState(newState: String, testMatrix: TestMatrix) {
-        state = newState
-        when (state) {
-            FINISHED -> finished(testMatrix)
+internal fun SavedMatrix.updateWithMatrix(newMatrix: TestMatrix): SavedMatrix =
+    if (needsUpdate(newMatrix)) updatedSavedMatrix(newMatrix)
+    else this
+
+private fun SavedMatrix.updatedSavedMatrix(newMatrix: TestMatrix): SavedMatrix {
+    var outcomeDetails = if (this.outcomeDetails.isNullOrEmpty()) "" else this.outcomeDetails
+    var outcome = this.outcome
+    var billableVirtualMinutes = this.billableVirtualMinutes
+    var billablePhysicalMinutes = this.billablePhysicalMinutes
+    if (this.state != newMatrix.state) {
```

Comment:
these `this` could be skipped :)

Reply:
No it cant :) if i do skip this, a recursion bug will occur when a badmatrix exception will occur. This is needed as the inital state of the object may differ from the incoming newmatrix state but if it doesnt differ there might only be minor updates to the data classes variable

Reply:
whoops made a mistake. Ignore the last comment. Ill remove "this"